Source/JavaScriptCore/runtime/VMTraps.cpp

```diff
@@ -257,31 +268,50 @@ class VMTraps::SignalSender final : public AutomaticThread {
         });
     }
 
-    ASCIILiteral name() const final
-    {
-        return "JSC VMTraps Signal Sender Thread"_s;
-    }
-
     VMTraps& traps() { return m_vm.traps(); }
 
-private:
-    PollResult poll(const AbstractLocker&) final
-    {
-        if (traps().m_isShuttingDown)
-            return PollResult::Stop;
 
-        if (!traps().needHandling(VMTraps::AsyncEvents))
-            return PollResult::Wait;
+    void notify(AbstractLocker&)
+    {
+        if (m_scheduled)
+            return;
+        m_scheduled = true;
+        queue().dispatch([protectedThis = Ref { *this }] {
+            protectedThis->work();
+        });
+    }
 
-        // We know that no trap could have been processed and re-added because we are holding the lock.
-        if (vmIsInactive(m_vm))
-            return PollResult::Wait;
-        return PollResult::Work;
+    bool isStopped(AbstractLocker&)
+    {
+        return !m_scheduled;
     }
 
-    WorkResult work() final
+private:
+    void work()
     {
         VM& vm = m_vm;
+        {
+            Locker locker { *m_lock };
+            ASSERT(m_scheduled);
+            if (traps().m_isShuttingDown) {
+                m_scheduled = false;
+                m_condition->notifyAll();
+                return;
```

Comment:
nit: This pattern is repeated many times, and all the places rely on doing each of these steps in order to have correctness.  If we ever need to add any more steps for correctness, it would be good to be able to automatically update get every place that needs it.  As such, I suggest making a lambda function to encapsulate this, and have all the places just call that method.  Maybe something like this?

```
auto workDone = [&] () {
    m_scheduled = false;
    m_condition->notifyAll(); // let work queue service next SignalSender if needed.
};
```
... and do the returns like:
```
    return workDone();
```

Reply:
Sounds good! Changed.